source/agora/consensus/Validation.d

```diff
@@ -631,6 +631,268 @@ unittest
     }
 }
 
+/// Test for transactions having no input or no output
+unittest
+{
+    import std.format;
+    Transaction[Hash] storage;
+    KeyPair key_pair = KeyPair.random;
+
+    // delegate for finding `Output`
+    scope findUTXO = (Hash hash, size_t index, out UTXOSetValue value)
+    {
+        if (auto tx = hash in storage)
+        {
+            if (index < tx.outputs.length)
+            {
+                value.unlock_height = 0;
+                value.type = TxType.Payment;
+                value.output = tx.outputs[index];
+                return true;
+            }
+        }
+        return false;
+    };
+
+    // Creates the transaction having no input
+    Transaction oneTx = Transaction(
+        TxType.Payment,
+        [],
+        [Output(Amount(50), key_pair.address)]
+    );
+    Hash oneHash = hashFull(oneTx);
+    storage[oneHash] = oneTx;
+
+    // Test for Payment transaction having no input
+    assert(!oneTx.isValid(findUTXO, 0), 
+        format("Tx having no input is not validated %s", oneTx));
+
+    // Creates a transaction having no output
+    Transaction firstTx = newCoinbaseTX(key_pair.address, Amount(100_1000));
+    Hash firstHash = hashFull(firstTx);
+    storage[firstHash] = firstTx;
+    firstTx.inputs[0].signature = key_pair.secret.sign(firstHash[]); 
+    Transaction secondTx = Transaction(
+        TxType.Payment,
+        [Input(firstHash, 0)],
+        []
+    );
+    Hash secondHash = hashFull(secondTx);
+    storage[secondHash] = secondTx;
+
+    // Test for Freeze transaction having no output
+    assert(!secondTx.isValid(findUTXO, 0), 
```

Comment:
I would use `invalidReason().cointains("No output")`.
You don't want to match the full message (it creates too much coupling which can be annoying), but it's very easy to have a test pass for the wrong reason.